presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java

```diff
@@ -0,0 +1,581 @@
+        }
+
+        private Set<ColumnConstraint> parseConstraints(TableHandle tableHandle, TupleDomain<ColumnHandle> constraint)
+        {
+            checkArgument(!constraint.isNone());
+            ImmutableSet.Builder<ColumnConstraint> columnConstraints = ImmutableSet.builder();
+            for (Map.Entry<ColumnHandle, Domain> entry : constraint.getDomains().get().entrySet()) {
+                ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableHandle, entry.getKey());
+                columnConstraints.add(new ColumnConstraint(columnMetadata.getName(), columnMetadata.getType().getTypeSignature(), parseDomain(entry.getValue())));
+            }
+            return columnConstraints.build();
+        }
+
+        private FormattedDomain parseDomain(Domain domain)
+        {
+            ImmutableSet.Builder<FormattedRange> formattedRanges = ImmutableSet.builder();
+            Type type = domain.getType();
+
+            domain.getValues().getValuesProcessor().consume(
+                    ranges -> formattedRanges.addAll(
+                            ranges.getOrderedRanges().stream()
+                                    .map(range -> new FormattedRange(formatMarker(range.getLow()), formatMarker(range.getHigh())))
+                                    .collect(toImmutableSet())),
+                    discreteValues -> formattedRanges.addAll(
+                            discreteValues.getValues().stream()
+                                    .map(value -> getVarcharValue(type, value))
+                                    .map(value -> new FormattedMarker(Optional.of(value), EXACTLY))
+                                    .map(marker -> new FormattedRange(marker, marker))
+                                    .collect(toImmutableSet())),
+                    allOrNone -> {
+                        throw new IllegalStateException("Unreachable AllOrNone consumer");
+                    });
+
+            return new FormattedDomain(domain.isNullAllowed(), formattedRanges.build());
+        }
+
+        private FormattedMarker formatMarker(Marker marker)
+        {
+            if (!marker.getValueBlock().isPresent()) {
+                return new FormattedMarker(Optional.empty(), marker.getBound());
+            }
+            return new FormattedMarker(Optional.of(getVarcharValue(marker.getType(), marker.getValue())), marker.getBound());
+        }
+
+        private String getVarcharValue(Type type, Object value)
+        {
+            if (type instanceof VarcharType) {
+                return ((Slice) value).toStringUtf8();
+            }
+            if (type instanceof TinyintType || type instanceof SmallintType || type instanceof IntegerType || type instanceof BigintType) {
+                return ((Long) value).toString();
+            }
+            throw new PrestoException(NOT_SUPPORTED, format("Unsupported data type in EXPLAIN (TYPE IO): %s", type.getDisplayName()));
```

Comment:
Thanks for this additional functionality, @shixuan-fan. Any thoughts around increasing support for additional types such as TIMESTAMP, DATE, etc?

Reply:
How about refactoring PlanPrinter.castToVarchar() so it can also be used by IOPlanPrinter?

Reply:
Casting an arbitrary type to varchar is okay for other plan printer because it is eventually for human reading. Yet for JSON format, the consumer is both machine and human, so we are conservative and using a whitelist strategy to make sure everything passes this conversion could surely be converted back (see https://github.com/prestodb/presto/pull/11049#discussion_r206356547). Of course we could add more type supports if needed :)

Reply:
The one we need support for now is TIMESTAMP (and probably DATE eventually). If the consumer of the JSON knows the type of the value, using PlanPrinter.castToVarchar() would work. How about wrapping PlanPrinter.castToVarchar() and throwing in the case of <UNREPRESENTABLE VALUE> (and other troublesome cases)? Or throwing in that case with a catch in the PlanPrinter usage?

Reply:
Because we are using whitelisting method, having extra dependency might not be good because we have to make the assumption about how `PlanPrinter.castToVarchar()` works. Probably it would be better to add new supported types in this method. cc @haozhun

Reply:
In our use case, we're only interested in the columns being filtered on (and also whether it's an equality or range comparison), not the constant value on the RHS. How about a config property (defaulting to false) that if set to true causes a constant string to be returned as a kind of marker (like "__?__") instead of throwing?